runtime/event_loop.cpp

```diff
@@ -39,53 +40,74 @@ bool EventLoop::cancel_async_task(api::Engine *engine, const int32_t id) {
 
 bool EventLoop::has_pending_async_tasks() { return !queue.get().tasks.empty(); }
 
-void EventLoop::incr_event_loop_lifetime() {
-  queue.get().lifetime_cnt++;
-}
+void EventLoop::incr_event_loop_interest() { queue.get().interest_cnt++; }
 
-void EventLoop::decr_event_loop_lifetime() {
-  MOZ_ASSERT(queue.get().lifetime_cnt > 0);
-  queue.get().lifetime_cnt--;
+void EventLoop::decr_event_loop_interest() {
+  MOZ_ASSERT(queue.get().interest_cnt > 0);
+  queue.get().interest_cnt--;
 }
 
-bool lifetime_complete() {
-  return queue.get().lifetime_cnt == 0;
-}
+inline bool interest_complete() { return queue.get().interest_cnt == 0; }
+
+inline void exit_event_loop() { queue.get().event_loop_running = false; }
 
 bool EventLoop::run_event_loop(api::Engine *engine, double total_compute) {
+  if (queue.get().event_loop_running) {
+    fprintf(stderr, "cannot run event loop as it is already running");
+    return false;
+  }
+  queue.get().event_loop_running = true;
   JSContext *cx = engine->cx();
 
   while (true) {
     while (js::HasJobsPending(cx)) {
       js::RunJobs(cx);
-      if (JS_IsExceptionPending(cx))
+      if (JS_IsExceptionPending(cx)) {
+        exit_event_loop();
         return false;
-      if (lifetime_complete())
+      }
+      if (interest_complete()) {
+        exit_event_loop();
         return true;
+      }
```

Comment:
Good point, I've updated the implementation to ensure a succesfull completion on no interest even if there are no async tasks, so that no interest always corresponds to "one tick".